# Objective

The main objective of this HR Analytics project is to explore and analyze the dataset for "Atlas Labs". The project aims to preprocess, and visualize the data using Tableau to create a clean, self-explanatory, and interactive dashboard for the HR team of Atlas Labs. The dashboard will provide valuable insights into employee attrition and the factors that impact attrition.

Data preprocessing has been performed in this Google Colab, including data cleaning, data wrangling, and anomaly detection. Ensuring data quality is crucial to obtaining reliable and accurate results from the analysis. As the saying goes, "*garbage in, garbage out*", meaning that the conclusions drawn from data analysis can only be as reliable as the data used. Hence, the project emphasizes the importance of using clean, accurate, and reliable data to derive meaningful and trustworthy insights.

By combining the power of data analysis and visualization, this HR Analytics project seeks to provide valuable and actionable insights that will contribute to fostering a positive work environment, optimizing talent management strategies, and ultimately improving employee satisfaction and retention at Atlas Labs.

# Data Preparation

In [92]:
# import pandas
import pandas as pd

In [93]:
# import datasets
employee = pd.read_csv('employee.csv', parse_dates=['HireDate'])
performance_rating = pd.read_csv('performance_rating.csv', delimiter=';', parse_dates=['ReviewDate'])
education_level = pd.read_csv('education_level.csv')

In [94]:
performance_rating.head()

PerformanceID EmployeeID ReviewDate EnvironmentSatisfaction JobSatisfaction  \
0          PR01  79F7-78EC 2013-01-02          Very Satisfied       Satisfied   
1          PR02  B61E-0F26 2013-01-03          Very Satisfied       Satisfied   
2          PR03  F5E3-48BB 2013-01-03                 Neutral       Satisfied   
3          PR04  0678-748A 2013-01-04          Very Satisfied         Neutral   
4          PR05  541F-3E19 2013-01-04          Very Satisfied    Dissatisfied   

  RelationshipSatisfaction  TrainingOpportunitiesWithinYear  \
0           Very Satisfied                                1   
1                Satisfied                                1   
2           Very Satisfied                                3   
3             Dissatisfied                                2   
4                  Neutral                                1   

   TrainingOpportunitiesTaken WorkLifeBalance           SelfRating  \
0                           0       Satisfied  Exceeds Expectation   
1                           3       Satisfied  Exceeds Expectation   
2                           2         Neutral     Above and Beyond   
3                           0    Dissatisfied    Meets Expectation   
4                           0       Satisfied  Exceeds Expectation   

         ManagerRating  
0  Exceeds Expectation  
1    Meets Expectation  
2  Exceeds Expectation  
3    Needs Improvement  
4    Meets Expectation

In [95]:
employee.head()

EmployeeID   FirstName LastName      Gender  Age BusinessTravel  \
0  3012-1A41    Leonelle    Simco      Female   30    Some Travel   
1  CBCB-9C9D     Leonerd    Aland        Male   38    Some Travel   
2  95D7-1CE9       Ahmed    Sykes        Male   43    Some Travel   
3  47A0-559B  Ermentrude   Berrie  Non-Binary   39    Some Travel   
4  42CC-040A       Stace   Savege      Female   29    Some Travel   

        Department  DistanceFromHome (KM) State                Ethnicity  ...  \
0            Sales                     27    IL                    White  ...   
1            Sales                     23    CA                    White  ...   
2  Human Resources                     29    CA  Asian or Asian American  ...   
3       Technology                     12    IL                    White  ...   
4  Human Resources                     29    CA                    White  ...   

   MaritalStatus  Salary StockOptionLevel OverTime   HireDate  Attrition  \
0       Divorced  102059                1       No 2012-01-03         No   
1         Single  157718                0      Yes 2012-01-04         No   
2        Married  309964                1       No 2012-01-04         No   
3        Married  293132                0       No 2012-01-05         No   
4         Single   49606                0       No 2012-01-05        Yes   

  YearsAtCompany YearsInMostRecentRole YearsSinceLastPromotion  \
0             10                     4                       9   
1             10                     6                      10   
2             10                     6                      10   
3             10                    10                      10   
4              6                     1                       1   

   YearsWithCurrManager  
0                     7  
1                     0  
2                     8  
3                     0  
4                     6  

[5 rows x 23 columns]

In [96]:
education_level

EducationLevelID            EducationLevel
0                 1  No Formal Qualifications
1                 2               High School
2                 3                 Bachelors
3                 4                   Masters
4                 5                 Doctorate

# Data Cleaning

## Checking Missing Value

In [97]:
performance_rating.isna().sum()

PerformanceID                      0
EmployeeID                         0
ReviewDate                         0
EnvironmentSatisfaction            0
JobSatisfaction                    0
RelationshipSatisfaction           0
TrainingOpportunitiesWithinYear    0
TrainingOpportunitiesTaken         0
WorkLifeBalance                    0
SelfRating                         0
ManagerRating                      0
dtype: int64

In [98]:
employee.isna().sum()

EmployeeID                 0
FirstName                  0
LastName                   0
Gender                     0
Age                        0
BusinessTravel             0
Department                 0
DistanceFromHome (KM)      0
State                      0
Ethnicity                  0
Education                  0
EducationField             0
JobRole                    0
MaritalStatus              0
Salary                     0
StockOptionLevel           0
OverTime                   0
HireDate                   0
Attrition                  0
YearsAtCompany             0
YearsInMostRecentRole      0
YearsSinceLastPromotion    0
YearsWithCurrManager       0
dtype: int64

However, upon inspection, it was found that all three datasets do not contain any missing values. This is a good sign, as it ensures the data's completeness and saves us from the need to handle missing data imputation.

## Duplicated Data

In [99]:
employee.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 23 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   EmployeeID               1470 non-null   object        
 1   FirstName                1470 non-null   object        
 2   LastName                 1470 non-null   object        
 3   Gender                   1470 non-null   object        
 4   Age                      1470 non-null   int64         
 5   BusinessTravel           1470 non-null   object        
 6   Department               1470 non-null   object        
 7   DistanceFromHome (KM)    1470 non-null   int64         
 8   State                    1470 non-null   object        
 9   Ethnicity                1470 non-null   object        
 10  Education                1470 non-null   int64         
 11  EducationField           1470 non-null   object        
 12  JobRole                  1470 non-

In [100]:
employee.nunique()

EmployeeID                 1468
FirstName                  1334
LastName                   1441
Gender                        4
Age                          34
BusinessTravel                3
Department                    3
DistanceFromHome (KM)        45
State                         3
Ethnicity                     7
Education                     5
EducationField                8
JobRole                      13
MaritalStatus                 3
Salary                     1455
StockOptionLevel              4
OverTime                      2
HireDate                   1192
Attrition                     2
YearsAtCompany               11
YearsInMostRecentRole        11
YearsSinceLastPromotion      11
YearsWithCurrManager         11
dtype: int64

In [101]:
employee.duplicated().sum()

0

Number of unique `EmployeeID` (1468) is lesser than the total of `EmployeeID` (1470). This means that there are `EmployeeID` duplicates. The `EmployeeID` duplicates turned out to be same value assigned to three different employees. I will assign a unique `EmployeeID` to these three employees after doing the table merge with `education_level` table and `performance_rating` table.

In [102]:
employee_duplicate = employee[employee.duplicated("EmployeeID", keep=False)]
employee_duplicate

EmployeeID FirstName  LastName  Gender  Age BusinessTravel  Department  \
545    0.00E+00    Jamesy  Barnardo    Male   26    Some Travel       Sales   
667    0.00E+00      Scot    Gamlen    Male   23    Some Travel       Sales   
1172   0.00E+00      Pepi     Clift  Female   21    Some Travel  Technology   

      DistanceFromHome (KM) State                  Ethnicity  ...  \
545                      16    CA  Black or African American  ...   
667                      26    NY            Native Hawaiian  ...   
1172                     39    NY                      White  ...   

      MaritalStatus  Salary StockOptionLevel OverTime   HireDate  Attrition  \
545          Single  113512                0       No 2015-12-25         No   
667          Single   58345                0      Yes 2017-01-19         No   
1172         Single   63201                0       No 2020-12-13         No   

     YearsAtCompany YearsInMostRecentRole YearsSinceLastPromotion  \
545               7                     3                       7   
667               5                     4                       4   
1172              2                     2                       2   

      YearsWithCurrManager  
545                      6  
667                      4  
1172                     2  

[3 rows x 23 columns]

# Data Wrangling

## Merge Table & Handling Duplicate

In [103]:
# merge the employee table with the education_level
emp_edu = employee.merge(education_level, left_on="Education", right_on="EducationLevelID")
emp_edu.head()

EmployeeID  FirstName  LastName  Gender  Age      BusinessTravel  \
0  3012-1A41   Leonelle     Simco  Female   30         Some Travel   
1  C0C7-A3EF      Hagen     Worge    Male   48           No Travel   
2  3F19-FC5A         Hy  Trevance    Male   35         Some Travel   
3  A2FB-2D5D     Lebbie     Poure  Female   43         Some Travel   
4  DEBE-1622  Bartolemo     Balma    Male   28  Frequent Traveller   

   Department  DistanceFromHome (KM) State Ethnicity  ...  StockOptionLevel  \
0       Sales                     27    IL     White  ...                 1   
1       Sales                      9    CA     White  ...                 0   
2       Sales                     29    NY     White  ...                 2   
3  Technology                     11    CA     White  ...                 1   
4  Technology                     22    CA     White  ...                 0   

  OverTime   HireDate Attrition  YearsAtCompany  YearsInMostRecentRole  \
0       No 2012-01-03        No              10                      4   
1       No 2012-02-12        No              10                     10   
2      Yes 2012-03-10        No              10                      7   
3       No 2012-03-17        No              10                      9   
4       No 2012-04-02       Yes               4                      1   

  YearsSinceLastPromotion YearsWithCurrManager EducationLevelID  \
0                       9                    7                5   
1                      10                    4                5   
2                       9                    9                5   
3                       9                    1                5   
4                       2                    1                5   

   EducationLevel  
0       Doctorate  
1       Doctorate  
2       Doctorate  
3       Doctorate  
4       Doctorate  

[5 rows x 25 columns]

In [104]:
# merge the emp_edu table with the performance_rating
merged = emp_edu.merge(performance_rating, on='EmployeeID')
merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6729 entries, 0 to 6728
Data columns (total 35 columns):
 #   Column                           Non-Null Count  Dtype         
---  ------                           --------------  -----         
 0   EmployeeID                       6729 non-null   object        
 1   FirstName                        6729 non-null   object        
 2   LastName                         6729 non-null   object        
 3   Gender                           6729 non-null   object        
 4   Age                              6729 non-null   int64         
 5   BusinessTravel                   6729 non-null   object        
 6   Department                       6729 non-null   object        
 7   DistanceFromHome (KM)            6729 non-null   int64         
 8   State                            6729 non-null   object        
 9   Ethnicity                        6729 non-null   object        
 10  Education                        6729 non-null   int64      

In [105]:
merged.sort_values(by='ReviewDate', ascending=False).head()

EmployeeID  FirstName  LastName  Gender  Age      BusinessTravel  \
6675  AAD7-8DDC     Simeon  Cosslett    Male   50           No Travel   
6407  E47D-E3C5      Kerry    Orrell    Male   26         Some Travel   
6183  211E-BC82      Britt  Ritchley  Female   28  Frequent Traveller   
3869  6CC9-81FC   Anatollo      Dake    Male   25           No Travel   
6003  ACA9-5DC2  Mireielle   Chstney  Female   46  Frequent Traveller   

      Department  DistanceFromHome (KM) State                  Ethnicity  ...  \
6675  Technology                     11    NY                      White  ...   
6407  Technology                     14    NY  Black or African American  ...   
6183  Technology                      6    CA                      White  ...   
3869  Technology                     43    CA                      White  ...   
6003       Sales                     14    NY                      White  ...   

      PerformanceID ReviewDate EnvironmentSatisfaction JobSatisfaction  \
6675         PR6709 2022-12-31               Satisfied  Very Satisfied   
6407         PR6705 2022-12-29               Satisfied       Satisfied   
6183         PR6704 2022-12-29                 Neutral         Neutral   
3869         PR6708 2022-12-29                 Neutral    Dissatisfied   
6003         PR6703 2022-12-29            Dissatisfied         Neutral   

      RelationshipSatisfaction  TrainingOpportunitiesWithinYear  \
6675              Dissatisfied                                1   
6407                 Satisfied                                3   
6183                   Neutral                                2   
3869            Very Satisfied                                3   
6003         Very Dissatisfied                                3   

     TrainingOpportunitiesTaken    WorkLifeBalance           SelfRating  \
6675                          2       Dissatisfied  Exceeds Expectation   
6407                          2            Neutral     Above and Beyond   
6183                          3  Very Dissatisfied  Exceeds Expectation   
3869                          0       Dissatisfied  Exceeds Expectation   
6003                          0          Satisfied     Above and Beyond   

            ManagerRating  
6675  Exceeds Expectation  
6407  Exceeds Expectation  
6183    Meets Expectation  
3869    Meets Expectation  
6003  Exceeds Expectation  

[5 rows x 35 columns]

After the data has been successfully merged, it is time to replace the duplicate `EmployeeID`. Because here we want to know the attrition in the company, the `EmployeeID` becomes very important where we have to ensure that each employee has their own ID.

In [106]:
merged[merged['EmployeeID'] == '0.00E+00']

EmployeeID FirstName  LastName  Gender  Age BusinessTravel  Department  \
5389   0.00E+00    Jamesy  Barnardo    Male   26    Some Travel       Sales   
5390   0.00E+00    Jamesy  Barnardo    Male   26    Some Travel       Sales   
5391   0.00E+00    Jamesy  Barnardo    Male   26    Some Travel       Sales   
5392   0.00E+00    Jamesy  Barnardo    Male   26    Some Travel       Sales   
5393   0.00E+00    Jamesy  Barnardo    Male   26    Some Travel       Sales   
5394   0.00E+00    Jamesy  Barnardo    Male   26    Some Travel       Sales   
5395   0.00E+00    Jamesy  Barnardo    Male   26    Some Travel       Sales   
5396   0.00E+00    Jamesy  Barnardo    Male   26    Some Travel       Sales   
5397   0.00E+00    Jamesy  Barnardo    Male   26    Some Travel       Sales   
5398   0.00E+00    Jamesy  Barnardo    Male   26    Some Travel       Sales   
5399   0.00E+00      Scot    Gamlen    Male   23    Some Travel       Sales   
5400   0.00E+00      Scot    Gamlen    Male   23    Some Travel       Sales   
5401   0.00E+00      Scot    Gamlen    Male   23    Some Travel       Sales   
5402   0.00E+00      Scot    Gamlen    Male   23    Some Travel       Sales   
5403   0.00E+00      Scot    Gamlen    Male   23    Some Travel       Sales   
5404   0.00E+00      Scot    Gamlen    Male   23    Some Travel       Sales   
5405   0.00E+00      Scot    Gamlen    Male   23    Some Travel       Sales   
5406   0.00E+00      Scot    Gamlen    Male   23    Some Travel       Sales   
5407   0.00E+00      Scot    Gamlen    Male   23    Some Travel       Sales   
5408   0.00E+00      Scot    Gamlen    Male   23    Some Travel       Sales   
5409   0.00E+00      Pepi     Clift  Female   21    Some Travel  Technology   
5410   0.00E+00      Pepi     Clift  Female   21    Some Travel  Technology   
5411   0.00E+00      Pepi     Clift  Female   21    Some Travel  Technology   
5412   0.00E+00      Pepi     Clift  Female   21    Some Travel  Technology   
5413   0.00E+00      Pepi     Clift  Female   21    Some Travel  Technology   
5414   0.00E+00      Pepi     Clift  Female   21    Some Travel  Technology   
5415   0.00E+00      Pepi     Clift  Female   21    Some Travel  Technology   
5416   0.00E+00      Pepi     Clift  Female   21    Some Travel  Technology   
5417   0.00E+00      Pepi     Clift  Female   21    Some Travel  Technology   
5418   0.00E+00      Pepi     Clift  Female   21    Some Travel  Technology   

      DistanceFromHome (KM) State                  Ethnicity  ...  \
5389                     16    CA  Black or African American  ...   
5390                     16    CA  Black or African American  ...   
5391                     16    CA  Black or African American  ...   
5392                     16    CA  Black or African American  ...   
5393                     16    CA  Black or African American  ...   
5394                     16    CA  Black or African American  ...   
5395                     16    CA  Black or African American  ...   
5396                     16    CA  Black or African American  ...   
5397                     16    CA  Black or African American  ...   
5398                     16    CA  Black or African American  ...   
5399                     26    NY            Native Hawaiian  ...   
5400                     26    NY            Native Hawaiian  ...   
5401                     26    NY            Native Hawaiian  ...   
5402                     26    NY            Native Hawaiian  ...   
5403                     26    NY            Native Hawaiian  ...   
5404                     26    NY            Native Hawaiian  ...   
5405                     26    NY            Native Hawaiian  ...   
5406                     26    NY            Native Hawaiian  ...   
5407                     26    NY            Native Hawaiian  ...   
5408                     26    NY            Native Hawaiian  ...   
5409                     39    NY                      White  ...   
5410                     39    NY    

This ID is associated with 3 employees, indicating the presence of duplicate `EmployeeID`. To address this issue, we will label this ID as "`UnknownID`".

In [107]:
merged.loc[(merged['FirstName'] == 'Jamesy') & (merged['LastName'] == 'Barnardo'), 'EmployeeID'] = 'UnknownID1'
merged.loc[(merged['FirstName'] == 'Scott') & (merged['LastName'] == 'Gamlen'), 'EmployeeID'] = 'UnknownID2'
merged.loc[(merged['FirstName'] == 'Pepi') & (merged['LastName'] == 'Clift'), 'EmployeeID'] = 'UnknownID3'

merged[merged['LastName']=='Clift']

EmployeeID FirstName LastName  Gender  Age BusinessTravel  Department  \
5409  UnknownID3      Pepi    Clift  Female   21    Some Travel  Technology   
5410  UnknownID3      Pepi    Clift  Female   21    Some Travel  Technology   
5411  UnknownID3      Pepi    Clift  Female   21    Some Travel  Technology   
5412  UnknownID3      Pepi    Clift  Female   21    Some Travel  Technology   
5413  UnknownID3      Pepi    Clift  Female   21    Some Travel  Technology   
5414  UnknownID3      Pepi    Clift  Female   21    Some Travel  Technology   
5415  UnknownID3      Pepi    Clift  Female   21    Some Travel  Technology   
5416  UnknownID3      Pepi    Clift  Female   21    Some Travel  Technology   
5417  UnknownID3      Pepi    Clift  Female   21    Some Travel  Technology   
5418  UnknownID3      Pepi    Clift  Female   21    Some Travel  Technology   

      DistanceFromHome (KM) State Ethnicity  ...  PerformanceID ReviewDate  \
5409                     39    NY     White  ...         PR1198 2016-07-15   
5410                     39    NY     White  ...         PR1776 2017-07-15   
5411                     39    NY     White  ...         PR2449 2018-07-15   
5412                     39    NY     White  ...         PR3030 2019-04-24   
5413                     39    NY     White  ...          PR366 2014-07-16   
5414                     39    NY     White  ...         PR3909 2020-04-23   
5415                     39    NY     White  ...         PR4901 2021-04-23   
5416                     39    NY     White  ...         PR5979 2022-04-23   
5417                     39    NY     White  ...         PR6400 2022-09-05   
5418                     39    NY     White  ...          PR729 2015-07-16   

     EnvironmentSatisfaction JobSatisfaction  RelationshipSatisfaction  \
5409                 Neutral         Neutral                 Satisfied   
5410          Very Satisfied    Dissatisfied                 Satisfied   
5411                 Neutral  Very Satisfied                 Satisfied   
5412               Satisfied  Very Satisfied              Dissatisfied   
5413               Satisfied    Dissatisfied                 Satisfied   
5414          Very Satisfied         Neutral            Very Satisfied   
5415                 Neutral       Satisfied                   Neutral   
5416               Satisfied         Neutral            Very Satisfied   
5417                 Neutral    Dissatisfied            Very Satisfied   
5418          Very Satisfied         Neutral                   Neutral   

      TrainingOpportunitiesWithinYear TrainingOpportunitiesTaken  \
5409                                1                          1   
5410                                2                          1   
5411                                2                          1   
5412                                3                          1   
5413                                2                          1   
5414                                1                          1   
5415                                3                          0   
5416                                1                          2   
5417                                1                          1   
5418                                2                          3   

        WorkLifeBalance           SelfRating        ManagerRating  
5409       Dissatisfied    Meets Expectation    Meets Expectation  
5410          Satisfied  Exceeds Expectation  Exceeds Expectation  
5411       Dissatisfied  Exceeds Expectation    Meets Expectation  
5412            Neutral     Above and Beyond     Above and Beyond  
5413       Dissatisfied  Exceeds Expectation  Exceeds Expectation  
5414     Very Satisfied    Meets Expectation    Meets Expectation  
5415  Very Dissatisfied    Meets Expectation    Needs Improvement  
5416          Satisfied    Meets Expectation    Needs Improvement  
5417          Satisfied     Above and Beyond  Exceeds Expectation  
5418          Satisfied    Meets 

## Ordinal Encoding

Ordinal encoding is performed on the data to convert categorical values with a meaningful order into numerical form. In the dataset, we have satisfaction levels like "`Very Dissatisfied`", "`Dissatisfied`", "`Neutral`", "`Satisfied`", and "`Very Satisfied`". By applying ordinal encoding, we transform these satisfaction levels into numeric values like 1, 2, 3, 4, and 5, respectively, based on their meaningful order. Also we can do it at "`SelfRating`" and "`ManagerRating`".

It ensures that the model or analysis we perform can properly interpret and utilize the satisfaction ratings as meaningful numerical data, ultimately helping us gain valuable insights and make informed decisions based on the ordinal relationship between the satisfaction levels.

In [108]:
# Mapping for satisfaction levels to ordinal values
satisfaction = {"Very Dissatisfied": 1,
                "Dissatisfied": 2,
                "Neutral": 3,
                "Satisfied": 4,
                "Very Satisfied": 5}

# Using the .replace() method to perform ordinal encoding on multiple columns
satisfaction_to_encode = ['EnvironmentSatisfaction', 'JobSatisfaction', 'RelationshipSatisfaction', 'WorkLifeBalance']
for col in satisfaction_to_encode:
    merged[col] = merged[col].replace(satisfaction)

In [109]:
# Before doing the mapping, it was discovered that there was an input error in the ManagerRating
merged['ManagerRating'].unique()

array(['Meets Expectation', 'Needs Improvement', 'Above and Beyond',
       'Exceeds Expectation', 'Meets Expnctation'], dtype=object)

In [110]:
# Change the value "Meets Expnctation" to "Meets Expectation" in the ManagerRating
merged['ManagerRating'] = merged['ManagerRating'].replace('Meets Expnctation', 'Meets Expectation')

In [111]:
# Mapping for rating levels to ordinal values
rating = {"Needs Improvement": 1,
          "Meets Expectation": 2,
          "Exceeds Expectation": 3,
          "Above and Beyond": 4}

# Using the .replace() method to perform ordinal encoding
rating_to_encode = ['SelfRating', 'ManagerRating']
for col in rating_to_encode:
    merged[col] = merged[col].replace(rating)

In [112]:
merged.head()

EmployeeID FirstName LastName  Gender  Age BusinessTravel Department  \
0  3012-1A41  Leonelle    Simco  Female   30    Some Travel      Sales   
1  3012-1A41  Leonelle    Simco  Female   30    Some Travel      Sales   
2  3012-1A41  Leonelle    Simco  Female   30    Some Travel      Sales   
3  3012-1A41  Leonelle    Simco  Female   30    Some Travel      Sales   
4  3012-1A41  Leonelle    Simco  Female   30    Some Travel      Sales   

   DistanceFromHome (KM) State Ethnicity  ...  PerformanceID ReviewDate  \
0                     27    IL     White  ...         PR1295 2016-10-30   
1                     27    IL     White  ...         PR1908 2017-10-30   
2                     27    IL     White  ...         PR2617 2018-10-30   
3                     27    IL     White  ...         PR3436 2019-10-30   
4                     27    IL     White  ...          PR411 2014-10-31   

  EnvironmentSatisfaction JobSatisfaction  RelationshipSatisfaction  \
0                       3               3                         2   
1                       4               4                         5   
2                       5               5                         4   
3                       1               3                         2   
4                       3               4                         2   

   TrainingOpportunitiesWithinYear TrainingOpportunitiesTaken WorkLifeBalance  \
0                                3                          0               4   
1                                3                          1               2   
2                                3                          0               4   
3                                3                          1               3   
4                                1                          0               3   

  SelfRating  ManagerRating  
0          2              2  
1          2              1  
2          4              4  
3          4              3  
4          3              2  

[5 rows x 35 columns]

Please note that this data is not a duplicate, but this data contains reviews that are carried out by every employee every year from the day they joined the company. For example, an employee who joined in year 2015 would have 7 reviews up till year 2022.

## Label Encoding

In [113]:
from sklearn.preprocessing import LabelEncoder

# Inisialisasi LabelEncoder
label_encoder = LabelEncoder()

# Melakukan Label Encoding pada kolom 'Attrition'
merged['Attrition'] = label_encoder.fit_transform(merged['Attrition'])
merged['OverTime'] = label_encoder.fit_transform(merged['OverTime'])
merged[['EmployeeID','Attrition', 'OverTime']].sample(5)

EmployeeID  Attrition  OverTime
5164  D926-DED9          0         0
1528  B98B-4689          1         0
3110  1C6C-3B67          0         0
2469  565B-C207          0         0
1710  FD95-9EEB          1         1

## HireDate vs ReviewDate

In [114]:
# checking whether there is a review_date which is done before the hire_date
checked = merged[merged['HireDate'] <= merged['ReviewDate']]
checked.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5578 entries, 0 to 6710
Data columns (total 35 columns):
 #   Column                           Non-Null Count  Dtype         
---  ------                           --------------  -----         
 0   EmployeeID                       5578 non-null   object        
 1   FirstName                        5578 non-null   object        
 2   LastName                         5578 non-null   object        
 3   Gender                           5578 non-null   object        
 4   Age                              5578 non-null   int64         
 5   BusinessTravel                   5578 non-null   object        
 6   Department                       5578 non-null   object        
 7   DistanceFromHome (KM)            5578 non-null   int64         
 8   State                            5578 non-null   object        
 9   Ethnicity                        5578 non-null   object        
 10  Education                        5578 non-null   int64      

5578 data obtained from a total of 6729 data, which means there are rows which `ReviewDate` is before `HireDate`. This could be errors because it is not likely a yearly review took place before the employee joined the company. These erroneous rows will not be used. For example, Caryl joined in year 2017 but his yearly review records are on years 2013 to 2017.

In [115]:
# Review conducted before hire
review_after_hire = merged[merged['HireDate'] >= merged['ReviewDate']]
review_after_hire[['EmployeeID', 'FirstName', 'LastName', 'HireDate', 'ReviewDate']].head()

EmployeeID FirstName  LastName   HireDate ReviewDate
149  7749-B277     Caryl  Roycroft 2017-04-25 2017-02-28
151  7749-B277     Caryl  Roycroft 2017-04-25 2014-03-01
155  7749-B277     Caryl  Roycroft 2017-04-25 2015-03-01
157  7749-B277     Caryl  Roycroft 2017-04-25 2013-03-01
158  7749-B277     Caryl  Roycroft 2017-04-25 2016-02-29

# Exploratory Data Analysis

In [116]:
checked.describe()

Age  DistanceFromHome (KM)    Education         Salary  \
count  5578.000000            5578.000000  5578.000000    5578.000000   
mean     31.916637              22.066332     2.851380  116573.198996   
std       7.758390              12.865684     1.029404   98945.513742   
min      18.000000               1.000000     1.000000   20387.000000   
25%      26.000000              12.000000     2.000000   48930.000000   
50%      29.000000              21.000000     3.000000   82036.500000   
75%      38.000000              33.000000     4.000000  145328.000000   
max      51.000000              45.000000     5.000000  547204.000000   

       StockOptionLevel     OverTime    Attrition  YearsAtCompany  \
count       5578.000000  5578.000000  5578.000000     5578.000000   
mean           0.754751     0.300645     0.222481        6.553603   
std            0.826591     0.458580     0.415950        2.929954   
min            0.000000     0.000000     0.000000        0.000000   
25%            0.000000     0.000000     0.000000        4.000000   
50%            1.000000     0.000000     0.000000        7.000000   
75%            1.000000     1.000000     0.000000        9.000000   
max            3.000000     1.000000     1.000000       10.000000   

       YearsInMostRecentRole  YearsSinceLastPromotion  YearsWithCurrManager  \
count            5578.000000              5578.000000           5578.000000   
mean                3.299390                 4.921657              3.259591   
std                 2.829759                 3.002204              2.821385   
min                 0.000000                 0.000000              0.000000   
25%                 1.000000                 3.000000              1.000000   
50%                 3.000000                 5.000000              3.000000   
75%                 5.000000                 7.000000              5.000000   
max                10.000000                10.000000             10.000000   

       EducationLevelID  EnvironmentSatisfaction  JobSatisfaction  \
count       5578.000000              5578.000000      5578.000000   
mean           2.851380                 3.854966         3.423629   
std            1.029404                 0.948529         1.160893   
min            1.000000                 1.000000         1.000000   
25%            2.000000                 3.000000         2.000000   
50%            3.000000                 4.000000         3.000000   
75%            4.000000                 5.000000         4.000000   
max            5.000000                 5.000000         5.000000   

       RelationshipSatisfaction  TrainingOpportunitiesWithinYear  \
count               5578.000000                      5578.000000   
mean                   3.415023                         2.013804   
std                    1.159985                         0.821743   
min                    1.000000                         1.000000   
25%                    2.000000                         1.000000   
50%                    3.000000                         2.000000   
75%                    4.000000                         3.000000   
max                    5.000000                         3.000000   

       TrainingOpportunitiesTaken  WorkLifeBalance   SelfRating  ManagerRating  
count                 5578.000000      5578.000000  5578.000000    5578.000000  
mean                     1.013804         3.411617     2.979204       2.469702  
std                      0.952139         1.148389     0.816451       0.959404  
min                      0.000000         1.000000     2.000000       1.000000  
25%                      0.000000         2.000000     2.000000       2.000000  
50%                      1.000000         3.000000     3.000000       2.000000  
75%                      2.000000         4.000000     4.000000       3.000000  
max                      3.000000         5.000000     4.000000       4.000000

In [117]:
categorical_columns = ['Gender', 'BusinessTravel', 'Department', 'State', 'Ethnicity', 'EducationField',
                       'JobRole', 'MaritalStatus', 'OverTime', 'Attrition', 'EducationLevel']
for column in categorical_columns:
    unique_values = checked[column].unique()
    print(f"{column}: {unique_values}")
    print()

Gender: ['Female' 'Male' 'Non-Binary' 'Prefer Not To Say']

BusinessTravel: ['Some Travel' 'No Travel' 'Frequent Traveller']

Department: ['Sales' 'Technology' 'Human Resources']

State: ['IL' 'CA' 'NY']

Ethnicity: ['White' 'Mixed or multiple ethnic groups' 'Black or African American'
 'Asian or Asian American' 'Other' 'American Indian or Alaska Native'
 'Native Hawaiian']

EducationField: ['Marketing' 'Economics' 'Information Systems' 'Computer Science'
 'Business Studies' 'Human Resources' 'Other' 'Technical Degree']

JobRole: ['Sales Executive' 'Analytics Manager' 'Machine Learning Engineer'
 'Data Scientist' 'Software Engineer' 'Engineering Manager' 'Manager'
 'Senior Software Engineer' 'HR Executive' 'HR Business Partner'
 'Sales Representative' 'Recruiter' 'HR Manager']

MaritalStatus: ['Divorced' 'Single' 'Married']

OverTime: [0 1]

Attrition: [0 1]

EducationLevel: ['Doctorate' 'Masters' 'Bachelors' 'High School'
 'No Formal Qualifications']



In [118]:
checked.columns

Index(['EmployeeID', 'FirstName', 'LastName', 'Gender', 'Age',
       'BusinessTravel', 'Department', 'DistanceFromHome (KM)', 'State',
       'Ethnicity', 'Education', 'EducationField', 'JobRole', 'MaritalStatus',
       'Salary', 'StockOptionLevel', 'OverTime', 'HireDate', 'Attrition',
       'YearsAtCompany', 'YearsInMostRecentRole', 'YearsSinceLastPromotion',
       'YearsWithCurrManager', 'EducationLevelID', 'EducationLevel',
       'PerformanceID', 'ReviewDate', 'EnvironmentSatisfaction',
       'JobSatisfaction', 'RelationshipSatisfaction',
       'TrainingOpportunitiesWithinYear', 'TrainingOpportunitiesTaken',
       'WorkLifeBalance', 'SelfRating', 'ManagerRating'],
      dtype='object')

In [119]:
# Drop unnecessary column
cleaned_df = checked.copy()
cleaned_df.drop(columns='EducationLevelID', inplace=True)

# Data Exporting & Conclusion

## Final Check

In [120]:
cleaned_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5578 entries, 0 to 6710
Data columns (total 34 columns):
 #   Column                           Non-Null Count  Dtype         
---  ------                           --------------  -----         
 0   EmployeeID                       5578 non-null   object        
 1   FirstName                        5578 non-null   object        
 2   LastName                         5578 non-null   object        
 3   Gender                           5578 non-null   object        
 4   Age                              5578 non-null   int64         
 5   BusinessTravel                   5578 non-null   object        
 6   Department                       5578 non-null   object        
 7   DistanceFromHome (KM)            5578 non-null   int64         
 8   State                            5578 non-null   object        
 9   Ethnicity                        5578 non-null   object        
 10  Education                        5578 non-null   int64      

In [121]:
cleaned_df.nunique()

EmployeeID                         1217
FirstName                          1122
LastName                           1197
Gender                                4
Age                                  34
BusinessTravel                        3
Department                            3
DistanceFromHome (KM)                45
State                                 3
Ethnicity                             7
Education                             5
EducationField                        8
JobRole                              13
MaritalStatus                         3
Salary                             1208
StockOptionLevel                      4
OverTime                              2
HireDate                            998
Attrition                             2
YearsAtCompany                       11
YearsInMostRecentRole                11
YearsSinceLastPromotion              11
YearsWithCurrManager                 11
EducationLevel                        5
PerformanceID                      5568


In [122]:
cleaned_df.head()

EmployeeID FirstName LastName  Gender  Age BusinessTravel Department  \
0  3012-1A41  Leonelle    Simco  Female   30    Some Travel      Sales   
1  3012-1A41  Leonelle    Simco  Female   30    Some Travel      Sales   
2  3012-1A41  Leonelle    Simco  Female   30    Some Travel      Sales   
3  3012-1A41  Leonelle    Simco  Female   30    Some Travel      Sales   
4  3012-1A41  Leonelle    Simco  Female   30    Some Travel      Sales   

   DistanceFromHome (KM) State Ethnicity  ...  PerformanceID ReviewDate  \
0                     27    IL     White  ...         PR1295 2016-10-30   
1                     27    IL     White  ...         PR1908 2017-10-30   
2                     27    IL     White  ...         PR2617 2018-10-30   
3                     27    IL     White  ...         PR3436 2019-10-30   
4                     27    IL     White  ...          PR411 2014-10-31   

  EnvironmentSatisfaction JobSatisfaction  RelationshipSatisfaction  \
0                       3               3                         2   
1                       4               4                         5   
2                       5               5                         4   
3                       1               3                         2   
4                       3               4                         2   

   TrainingOpportunitiesWithinYear  TrainingOpportunitiesTaken  \
0                                3                           0   
1                                3                           1   
2                                3                           0   
3                                3                           1   
4                                1                           0   

  WorkLifeBalance  SelfRating  ManagerRating  
0               4           2              2  
1               2           2              1  
2               4           4              4  
3               3           4              3  
4               3           3              2  

[5 rows x 34 columns]

### Create a new dataframe

In the `cleaned_df` there are duplicates `EmployeeID`, also within the other columns, therefore I will create a new dataframe namely `unique_df`that only contains unique `EmployeeID`, which aims to facilitate analysis that focuses on Attrition employee, while the `cleaned_df` is more focused on each employees' ratings in each period of the year.

In [123]:
# Sort and drop duplicates based on 'EmployeeID' and keep only the latest date
cleaned_df.sort_values(by='ReviewDate', ascending=False, inplace=True)
unique_df = cleaned_df.drop_duplicates(subset='EmployeeID')

unique_df.nunique()

EmployeeID                         1217
FirstName                          1122
LastName                           1197
Gender                                4
Age                                  34
BusinessTravel                        3
Department                            3
DistanceFromHome (KM)                45
State                                 3
Ethnicity                             7
Education                             5
EducationField                        8
JobRole                              13
MaritalStatus                         3
Salary                             1208
StockOptionLevel                      4
OverTime                              2
HireDate                            998
Attrition                             2
YearsAtCompany                       11
YearsInMostRecentRole                11
YearsSinceLastPromotion              11
YearsWithCurrManager                 11
EducationLevel                        5
PerformanceID                      1215


In [124]:
unique_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1217 entries, 6675 to 1058
Data columns (total 34 columns):
 #   Column                           Non-Null Count  Dtype         
---  ------                           --------------  -----         
 0   EmployeeID                       1217 non-null   object        
 1   FirstName                        1217 non-null   object        
 2   LastName                         1217 non-null   object        
 3   Gender                           1217 non-null   object        
 4   Age                              1217 non-null   int64         
 5   BusinessTravel                   1217 non-null   object        
 6   Department                       1217 non-null   object        
 7   DistanceFromHome (KM)            1217 non-null   int64         
 8   State                            1217 non-null   object        
 9   Ethnicity                        1217 non-null   object        
 10  Education                        1217 non-null   int64   

In [125]:
unique_df.head()

EmployeeID  FirstName  LastName  Gender  Age      BusinessTravel  \
6675  AAD7-8DDC     Simeon  Cosslett    Male   50           No Travel   
3869  6CC9-81FC   Anatollo      Dake    Male   25           No Travel   
6003  ACA9-5DC2  Mireielle   Chstney  Female   46  Frequent Traveller   
6407  E47D-E3C5      Kerry    Orrell    Male   26         Some Travel   
4755  74F2-5C55    Skipper   Sambles    Male   47         Some Travel   

      Department  DistanceFromHome (KM) State                  Ethnicity  ...  \
6675  Technology                     11    NY                      White  ...   
3869  Technology                     43    CA                      White  ...   
6003       Sales                     14    NY                      White  ...   
6407  Technology                     14    NY  Black or African American  ...   
4755  Technology                     40    NY                      White  ...   

      PerformanceID ReviewDate EnvironmentSatisfaction JobSatisfaction  \
6675         PR6709 2022-12-31                       4               5   
3869         PR6708 2022-12-29                       3               2   
6003         PR6703 2022-12-29                       2               3   
6407         PR6705 2022-12-29                       4               4   
4755         PR6702 2022-12-29                       4               2   

      RelationshipSatisfaction  TrainingOpportunitiesWithinYear  \
6675                         2                                1   
3869                         5                                3   
6003                         1                                3   
6407                         4                                3   
4755                         4                                2   

      TrainingOpportunitiesTaken WorkLifeBalance  SelfRating  ManagerRating  
6675                           2               2           3              3  
3869                           0               2           3              2  
6003                           0               4           4              3  
6407                           2               3           4              3  
4755                           3               5           2              2  

[5 rows x 34 columns]

Because the data already has unique `EmployeeID`, we can see the proportion of employees who are attrition and not more easily.

In [130]:
unique_df[unique_df["Attrition"] == 0].count()

EmployeeID                         993
FirstName                          993
LastName                           993
Gender                             993
Age                                993
BusinessTravel                     993
Department                         993
DistanceFromHome (KM)              993
State                              993
Ethnicity                          993
Education                          993
EducationField                     993
JobRole                            993
MaritalStatus                      993
Salary                             993
StockOptionLevel                   993
OverTime                           993
HireDate                           993
Attrition                          993
YearsAtCompany                     993
YearsInMostRecentRole              993
YearsSinceLastPromotion            993
YearsWithCurrManager               993
EducationLevel                     993
PerformanceID                      993
ReviewDate               

In [131]:
unique_df[unique_df["Attrition"] == 1].count()

EmployeeID                         224
FirstName                          224
LastName                           224
Gender                             224
Age                                224
BusinessTravel                     224
Department                         224
DistanceFromHome (KM)              224
State                              224
Ethnicity                          224
Education                          224
EducationField                     224
JobRole                            224
MaritalStatus                      224
Salary                             224
StockOptionLevel                   224
OverTime                           224
HireDate                           224
Attrition                          224
YearsAtCompany                     224
YearsInMostRecentRole              224
YearsSinceLastPromotion            224
YearsWithCurrManager               224
EducationLevel                     224
PerformanceID                      224
ReviewDate               

In [134]:
# Calculate the percentage of employees who has "Attrition" (meaning they have left the company)
len(unique_df[unique_df["Attrition"] == 1]) / len(unique_df)*100

18.405916187345934

## Exporting Cleaned Data

In [129]:
# Export two dataframes (`cleaned_df` and `unique_df`) as csv file for data visualization in Tableau.
cleaned_df.to_csv('cleaned_rating.csv', index=False)
unique_df.to_csv('cleaned_employee.csv', index=False)